dotCMS/src/main/java/com/dotcms/security/secret/ServiceIntegrationAPIImpl.java

```diff
@@ -0,0 +1,554 @@
+                        String.format("Invalid attempt to override an existing file named '%s'.",
+                                incomingFile.getName()));
+            }
+
+            ymlMapper.writeValue(incomingFile, serviceDescriptor);
+
+            invalidateCache();
+        }
+    }
+
+    @Override
+    public void removeServiceIntegration(final String serviceKey, final User user,
+            boolean removeDescriptor)
+            throws DotSecurityException, DotDataException {
+        if (userDoesNotHaveAccess(user)) {
+            throw new DotSecurityException(String.format(
+                    "Invalid attempt to delete a service descriptors performed by user with id `%s`.",
+                    user.getUserId()));
+        }
+        final String serviceKeyLC = serviceKey.toLowerCase();
+        final ServiceDescriptorMeta serviceDescriptorMeta = getServiceDescriptorMap(user).get(serviceKeyLC);
+        if (null == serviceDescriptorMeta) {
+            throw new DotDataException(String.format("The requested descriptor `%s` does not exist.",serviceKey));
+        }else{
+            // if we succeed trying to find the descriptor.
+            // Lets get all the service-keys and organized by host-id
+            final Map<String, Set<String>> serviceKeysByHost = serviceKeysByHost();
+            for (final Entry<String, Set<String>> entry : serviceKeysByHost.entrySet()) {
+                final String hostId = entry.getKey();
+                final Set<String> serviceKeys = entry.getValue();
+                //Now we need to verify the service-key has a configuration for this host.
+                if(serviceKeys.contains(serviceKeyLC)){
+                    //And if it does we attempt to delete all the secrets.
+                    final Host host = hostAPI.find(hostId, user, false);
+                    deleteSecrets(serviceKey, host, user);
+                }
+            }
+            if(removeDescriptor) {
+                final String fileName = serviceDescriptorMeta.getFileName();
+                //Now we need to remove the file it self.
+                final String ymlFilesPath = getServiceDescriptorDirectory();
+                final Path file = Paths.get(ymlFilesPath + File.separator + fileName).normalize();
+                if (!file.toFile().exists()) {
+                    throw new DotDataException(
+                            String.format(" File with path `%s` does not exist. ", file));
+                }
+                try {
+                    Logger.debug(ServiceIntegrationAPIImpl.class, () -> String
+                            .format(" Failed attempt to delete file with path `%s` ", file));
+                    Files.delete(file);
+                } catch (IOException e) {
+                    throw new DotDataException(e);
+                }
+            }
+            invalidateCache();
+        }
+    }
+
+    private synchronized void invalidateCache(){
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use block level rather than method level synchronization](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4623512)